Detectors/TOF/calibration/src/TOFCalibCollector.cxx

```diff
@@ -0,0 +1,147 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "TOFCalibration/TOFCalibCollector.h"
+#include "Framework/Logger.h"
+#include <cassert>
+#include <iostream>
+#include <sstream>
+#include <TStopwatch.h>
+
+namespace o2
+{
+namespace tof
+{
+
+using Slot = o2::calibration::TimeSlot<o2::tof::TOFCalibInfoSlot>;
+
+//_____________________________________________
+void TOFCalibInfoSlot::fill(const gsl::span<const o2::dataformats::CalibInfoTOF> data)
+{
+  // fill container
+  // we do not apply any calibration at this stage, it will be applied when we
+  // process the data before filling the CCDB in the separate process
+
+  for (int i = data.size(); i--;) {
+    auto ch = data[i].getTOFChIndex();
+    auto offset = std::accumulate(mEntriesSlot.begin(), mEntriesSlot.begin() + ch, 0);
```

Comment:
Since "ch" is coming randomly, the only other possibility is to cache the number of entries before the first of every channel. What do you think?

Reply:
what is the typical data size? If it is large, you are basically doing N*N loop.
I would build the indices of ordered channel entries in data, then loop over increasing channels. This has also advantage of avoiding jumps. 
Smt.like:
````
int nd = data.size();
std::vector<int> ord(nd);
std::iota(ord.begin(), ord.end(), 0);
std::sort(ord.begin(), ord.end(), [&data](int i, int j) { return data[i].getTOFChIndex() < data[j].getTOFChIndex(); });
int chPrev=0, int offsPrev = 0;
for (int i=0;i<nd;i++) {
   const auto& dti = data[ord[i]];
   ch = dti.getTOFChIndex();
   auto offset = offsPref;
   if (ch>chPrev) {
      offset +=  std::accumulate(mEntriesSlot.begin() + chPrev, mEntriesSlot.begin() + ch, 0);
  }
  offsPrev = offset;
  chPrev = ch;
  ....
}

Reply:
The data size is typically 100 tracks per event. If the TF is ~12 ms, with 50 kHz, then we have ~60000 tracks per TF. 
Indeed ordering before processing is better.

Reply:
Cannot I just order data with the same sorting function, and access data in order, instead of building the vector of indices with which to access data?

Reply:
no, since the data is a const span. But even if it would be a vector: sorting means swapping (2 copies) of elements, in average n*log(n) times. For vectors on non-trivial elements this will be too expensive.

Reply:
Ok, then I will take your suggestion.

Reply:
Done